source/source_integration_test.go

```diff
@@ -0,0 +1,430 @@
+	defer cancel()
+
+	src := NewSource()
+
+	err = src.Configure(ctx, cfg)
+	is.NoErr(err)
+
+	err = src.Open(ctx, nil)
+	is.True(err != nil)
+
+	cancel()
+}
+
+func TestSource_Read_EmptyTable(t *testing.T) {
+	t.Parallel()
+
+	var (
+		ctx = context.Background()
+		cfg = prepareConfig(t)
+		is  = is.New(t)
+	)
+
+	repo, err := repository.New(cfg[config.URL])
+	is.NoErr(err)
+	defer repo.Close()
+
+	err = createTable(repo, cfg[config.Table])
+	is.NoErr(err)
+
+	defer func() {
+		err = dropTables(repo, cfg[config.Table])
+		is.NoErr(err)
+	}()
+
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	src := NewSource()
+
+	err = src.Configure(ctx, cfg)
+	is.NoErr(err)
+
+	err = src.Open(ctx, nil)
+	is.NoErr(err)
+
+	_, err = src.Read(ctx)
+	is.Equal(err, sdk.ErrBackoffRetry)
+
+	cancel()
+
+	err = src.Teardown(context.Background())
+	is.NoErr(err)
+}
+
+func TestSource_Snapshot_Read(t *testing.T) {
+	t.Parallel()
+
+	var (
+		ctx = context.Background()
+		cfg = prepareConfig(t)
+		is  = is.New(t)
+	)
+
+	repo, err := repository.New(cfg[config.URL])
+	is.NoErr(err)
+	defer repo.Close()
+
+	err = createTable(repo, cfg[config.Table])
+	is.NoErr(err)
+
+	defer func() {
+		err = dropTables(repo, cfg[config.Table])
+		is.NoErr(err)
+	}()
+
+	// insert snapshot data
+	err = insertSnapshotData(repo, cfg[config.Table])
+	is.NoErr(err)
+
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	src := NewSource()
+
+	err = src.Configure(ctx, cfg)
+	is.NoErr(err)
+
+	err = src.Open(ctx, nil)
+	is.NoErr(err)
+
+	// read records
+	record, err := src.Read(ctx)
+	is.NoErr(err)
+	is.Equal(string(record.Position), `{"mode":"snapshot","last_processed_val":1}`)
+	is.Equal(record.Operation, sdk.OperationSnapshot)
+	is.Equal(record.Key, sdk.StructuredData(map[string]interface{}{"ID": 1}))
+	is.Equal(strings.ReplaceAll(string(record.Payload.After.Bytes()), " ", ""),
+		`{"AGE":42,"ID":1,"IS_ACTIVE":true,"NAME":"John"}`)
+
+	record, err = src.Read(ctx)
+	is.NoErr(err)
+	is.Equal(string(record.Position), `{"mode":"snapshot","last_processed_val":2}`)
+	is.Equal(record.Operation, sdk.OperationSnapshot)
+	is.Equal(record.Key, sdk.StructuredData(map[string]interface{}{"ID": 2}))
+	is.Equal(strings.ReplaceAll(string(record.Payload.After.Bytes()), " ", ""),
+		`{"AGE":12,"ID":2,"IS_ACTIVE":false,"NAME":"Jane"}`)
+
+	cancel()
+
+	err = src.Teardown(context.Background())
+	is.NoErr(err)
+
+	ctx, cancel = context.WithCancel(context.Background())
+	defer cancel()
+
+	src = NewSource()
+
+	err = src.Configure(ctx, cfg)
+	is.NoErr(err)
+
+	// open source with no nil position
+	err = src.Open(ctx, record.Position)
+	is.NoErr(err)
+
+	record, err = src.Read(ctx)
+	is.NoErr(err)
+	is.Equal(string(record.Position), `{"mode":"snapshot","last_processed_val":3}`)
+	is.Equal(record.Operation, sdk.OperationSnapshot)
+	is.Equal(record.Key, sdk.StructuredData(map[string]interface{}{"ID": 3}))
+	is.Equal(strings.ReplaceAll(string(record.Payload.After.Bytes()), " ", ""),
```

Comment:
This library might make it easier to compare JSON strings: https://pkg.go.dev/github.com/stretchr/testify/require#JSONEq

Reply:
Oh, it is my bad.
I used to use the CHAR type and it returned a string of names with spaces. 
So I decided to not use `https://github.com/stretchr/testify` if you don't mind :)

Reply:
Even better!